server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -88,31 +95,38 @@ public void afterRefresh(boolean didRefresh) {
                         this.remoteDirectory.init();
                     }
                     try {
-                        String lastCommittedLocalSegmentFileName = SegmentInfos.getLastCommitSegmentsFileName(storeDirectory);
-                        if (!remoteDirectory.containsFile(
-                            lastCommittedLocalSegmentFileName,
-                            getChecksumOfLocalFile(lastCommittedLocalSegmentFileName)
-                        )) {
+                        // if a new segments_N file is present in local that is not uploaded to remote store yet, it
+                        // is considered as a first refresh post commit. A cleanup of stale commit files is triggered.
+                        // Ideally, we want this to be done in async flow. (GitHub issue #4315)
+                        if (isRefreshAfterCommit()) {
                             deleteStaleCommits();
                         }
+                        String segmentInfoSnapshotFilename = null;
                         try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
                             SegmentInfos segmentInfos = segmentInfosGatedCloseable.get();
+
                             Collection<String> refreshedLocalFiles = segmentInfos.files(true);
 
                             List<String> segmentInfosFiles = refreshedLocalFiles.stream()
                                 .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
                                 .collect(Collectors.toList());
                             Optional<String> latestSegmentInfos = segmentInfosFiles.stream()
-                                .max(Comparator.comparingLong(IndexFileNames::parseGeneration));
+                                .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
 
                             if (latestSegmentInfos.isPresent()) {
+                                Set<String> segmentFilesFromSnapshot = new HashSet<>(refreshedLocalFiles);
                                 refreshedLocalFiles.addAll(SegmentInfos.readCommit(storeDirectory, latestSegmentInfos.get()).files(true));
                                 segmentInfosFiles.stream()
                                     .filter(file -> !file.equals(latestSegmentInfos.get()))
                                     .forEach(refreshedLocalFiles::remove);
 
                                 boolean uploadStatus = uploadNewSegments(refreshedLocalFiles);
                                 if (uploadStatus) {
+                                    if (segmentFilesFromSnapshot.equals(new HashSet<>(refreshedLocalFiles))) {
```

Comment:
As we discussed  :

1. We should try to root cause why `indexShard.getSegmentInfosSnapshot()` doesn't give all the segment files correctly . There will be other side effects as well for this incorrect result. 
1. In case, we are not able to fix `indexShard.getSegmentInfosSnapshot()` , we should only upload metadata when we are able to upload `segmentFilesFromSnapshot` .  This is because if we upload metadata w/o `SegmentInfosSnapshot` , that data is useless . It even renders the previous uploaded refresh useless as we refer to only the latest metadata . And we would then need to fallback to last commit.